s3torchconnector/tst/unit/test_s3reader.py

```diff
@@ -301,3 +303,110 @@ def test_s3reader_writes_size_after_read_all_explicit(stream: List[bytes]):
     assert s3reader.read(1) == b""
     # Once we've read past the end, we know how big the file is
     assert s3reader._size == total_length
+
+
+@given(
+    lists(binary(min_size=2, max_size=3), min_size=0, max_size=3),
+    integers(min_value=0, max_value=1),
+)
+def test_s3reader_readinto_buffer_smaller_then_chunks(
+    stream: List[bytes], buf_size: int
+):
+    s3reader = S3Reader(TEST_BUCKET, TEST_KEY, lambda: None, lambda: iter(stream))
+    assert s3reader._size is None
+    total_length = sum(map(len, stream))
+    buf = memoryview(bytearray(buf_size))
+    # We're able to read all the available data or the data that can be accommodated in buf
+    if buf_size > 0 and total_length > 0:
+        assert s3reader.readinto(buf) == buf_size
+        assert s3reader.tell() == buf_size
+        # We haven't reached the end yet
+        assert s3reader._size is None
+        # confirm that read data is the same as in source
+        assert buf[:buf_size] == (b"".join(stream))[:buf_size]
+    else:
+        assert s3reader.readinto(buf) == 0
+        assert s3reader.tell() == 0
+
+
+@given(
+    lists(binary(min_size=2, max_size=3), min_size=1, max_size=3),
+    integers(min_value=3, max_value=10),
+)
+def test_s3reader_readinto_buffer_bigger_then_chunks(
+    stream: List[bytes], buf_size: int
+):
+    s3reader = S3Reader(TEST_BUCKET, TEST_KEY, lambda: None, lambda: iter(stream))
+    assert s3reader._size is None
+    total_length = sum(map(len, stream))
+    buf = memoryview(bytearray(buf_size))
+    should_read_bytes_in_first_pass = min(buf_size, total_length)
+    # We're able to read all the available data or the data that can be accommodated in buf
+    assert s3reader.readinto(buf) == should_read_bytes_in_first_pass
+    assert s3reader.tell() == should_read_bytes_in_first_pass
+    all_data = b"".join(stream)
+    # confirm that read data is the same as in source
+    assert (
+        buf[:should_read_bytes_in_first_pass]
+        == all_data[:should_read_bytes_in_first_pass]
+    )
+    if total_length < buf_size:
+        assert s3reader._size == total_length
+
+    should_read_bytes_in_second_pass = max(
+        min(buf_size, total_length - should_read_bytes_in_first_pass), 0
+    )
+    if should_read_bytes_in_second_pass > 0:
+        # We're able to read all the available data or the data that can be accommodated in buf
+        assert s3reader.readinto(buf) == should_read_bytes_in_second_pass
+        total_read = should_read_bytes_in_first_pass + should_read_bytes_in_second_pass
+        assert s3reader.tell() == total_read
+        # confirm that read data is the same as in source
+        assert (
+            buf[:should_read_bytes_in_second_pass]
+            == all_data[should_read_bytes_in_first_pass:total_read]
+        )
+        if total_length < total_read:
+            assert s3reader._size == total_read
+
```

Comment:
Could we have a simple test that verifies `readinto` with a buf with length equal or greater than of the total object always reads the entire thing?